### PR TITLE
[PyROOT] Fix for AsNumpy when column is empty

### PR DESCRIPTION
**Pull request by cwiel**
Currently the `RDataFrame.AsNumpy` function can not handle empty columns.

This minimal example produces an `ValueError: setting an array element with a sequence.`

```
def fill_tree(treeName, fileName):
    ROOT.gRandom.SetSeed(1)
    tdf = ROOT.ROOT.RDataFrame(100)
    tdf.Define("b1", "2.")

treeName = 'tree'
fileName = treeName+".root"
fill_tree(treeName, treeName+".root")
base = ROOT.ROOT.RDataFrame(treeName, fileName)

base.Filter("b1>3").AsNumpy(columns=["b1"])  
```

```
---------------------------------------
File "ROOT.py", line 441, in _RDataFrameAsNumpy
    tmp = numpy.array(cpp_reference) # This adopts the memory of the C++ object.
ValueError: setting an array element with a sequence.
```

### CODE REVIEW

File: bindings/pyroot/ROOT.py [hunk 1]
```diff
@@ -436,6 +436,8 @@ def __array_finalize__(self, obj):
     for column in columns:
         cpp_reference = result_ptrs[column].GetValue()
         if hasattr(cpp_reference, "__array_interface__"):
+            if cpp_reference.empty():
+              cpp_reference = []
```
Comment (etejedor) [inline bindings/pyroot/ROOT.py]: Thank you! Is the indentation correct here?

**Comment by stwunsch** [comment]
Uh, thank you alot for your contribution! This corner-case should be added to the tests as well. This would go into `bindings/pyroot_experimental/PyROOT/test/array_interface.py`. Could you add a test there?

**Comment by stwunsch** [comment]
Wait, I was too quick. This fix should go into the adoption mechanism through the array interface and not into `RDataFrame.AsNumpy`. The minimal reproducer is the following:

```python
import ROOT
import numpy
x = ROOT.std.vector("float")()
npy = numpy.asnumpy(x)
```

And then, I'm not sure what value the `data` field should take and what is consistent with what numpy does for empty arrays (which do not have any data?)

**Comment by stwunsch** [comment]
Hm, it looks like this:

```python
import numpy
x = numpy.empty(0)
x.__array_interface__
# {'descr': [('', '<f8')], 'strides': None, 'shape': (0,), 'version': 3, 'typestr': '<f8', 'data': (25479120, False)}
```
```

**Comment by cwiel** [comment]
I don't know, if I am able to fix it in the array interface.
Where is this array interface actually defined and where is the adoption mechanism implemented?

**Comment by stwunsch** [comment]
If you like to have a look into this, I would be happy to guide you. But feel free to tell me if you have to do other stuff, then I'll take over.

The array interface is added in `ROOT.py:243`, however no need to change something there. Following the code path, you can look for `AddArrayInterface` in `bindings/pyroot/src/Pythonize.cxx`. Still no need to fix something there ;) Finally, you end up in `STLVectorArrayInterface` and `RVecArrayInterface` in as well in `Pythonize.cxx`. There, the C++ object (a `std::vector` in your case) is read out and the `__array_interface__` dict is added to the Python object, which carries all information needed by numpy to adopt the memory (see [here](https://docs.scipy.org/doc/numpy/reference/arrays.interface.html)).

In short: You would need to cover the case of `vector.size() == 0` in `STLVectorArrayInterface` properly.

Following code tells me, that the size field is already set properly, but it seems that numpy fails with the `"data": (0, False)` set in the `__array_interface__`.

```python
>>> import ROOT
>>> x = ROOT.std.vector("float")()
>>> x.__array_interface__
{'shape': (0L,), 'version': 3, 'data': (0, False), 'typestr': '<f4'}
```

So what is the correct thing to do there that `numpy.asarray` returns an empty numpy array?

**Comment by cwiel** [comment]
Ok the issue seems to be the `0` in the `'data'` entry of the dict. 

probably this is interpreted as None if it is zero and numpy tries to read the data trough the buffer interface of the object.

>  If this key is not present (or returns None), then memory sharing will be done through the buffer interface of the object itself

```python
import numpy
class empty_test:
	__array_interface__ = \
     {'shape': (0L,), 
     'version': 3, 
     'data': (1, False), 
     'typestr': '<f4'}

print(numpy.asarray(empty_test()))
```

so something like

```c++
if(cobj)->empty()){
      PyDict_SetItemString(dict, "data",
                           PyTuple_Pack(2, PyLong_FromLong(reinterpret_cast<long>(-1)), Py_False));
}
```

in the Pythonization.cxx
should fix this.

**Comment by stwunsch** [comment]
I can confirm that this fixes the issue. Even though I don't understand from the numpy side why `-1` is a better pointer than `0` :P

**Comment by stwunsch** [comment]
Ok, I've just spotted some other issues which did not yet show up as a bug but could lead to similar behaviours (for example, the type of the casted address is only long but should by unsigned long long to cover the full 64 bit address space).

I'm very grateful for the time you put into this and finding out what is actually going wrong. I'm gonna spend this afternoon in revamping and hardening the adoption mechanism. As soon as I've put all fixes in, I'll report back!

**Comment by stwunsch** [comment]
I've put the fix in the PR #3535.
